File: Common Algorithm/Tree Traversal/Depth First Search/InOrder Traversal/InOrder.java
import java.util.ArrayList;
import java.util.Stack;

public class InOrder {
    public static void main(String[] args) {
        TreeNode treeNode = new TreeNode(1, 
            new TreeNode(2, 
                new TreeNode(4, null, null), 
                new TreeNode(5, null, null)),
            new TreeNode(3,
                new TreeNode(6, null, null),
                null));

        inorderTraverse(treeNode);
    }

    static void inorderTraverse(TreeNode root) {
        if (root == null) return;

        Stack<TreeNode> stack = new Stack<>();
        ArrayList<Integer> result = new ArrayList<>();

        while (root != null || stack.size() != 0) {
            while (root != null) {
                stack.push(root);
                root = root.left;
            }

            root = stack.pop();
            result.add(root.val);
            root = root.right;
        }

        System.out.println(result);
    }
    
    static class TreeNode {
        public TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }

        public int val;
        public TreeNode left;
        public TreeNode right;
    }
}

File: Common Algorithm/Tree Traversal/Depth First Search/InOrder Traversal/inorder.kt


import java.util.Stack

class TreeNode(var `val`: Int, var left: TreeNode? = null, var right: TreeNode? = null)

fun inorderTraversal(root: TreeNode?): List<Int> {
    val result = mutableListOf<Int>()
    val stack = Stack<TreeNode>()

    var current: TreeNode? = root

    while (current != null || !stack.isEmpty()) {
        while (current != null) {
            stack.push(current)
            current = current.left
        }

        current = stack.pop()
        result.add(current.`val`)
        current = current.right
    }

    return result
}

fun main() {
    // Create a binary tree
    //        1
    //      /   \
    //     2     3
    //    / \   / \
    //   4   5 6   7
    val root = TreeNode(1,
        TreeNode(2,
            TreeNode(4),
            TreeNode(5)),
        TreeNode(3,
            TreeNode(6),
            TreeNode(7)))

    // Traverse the tree in inorder and print the values
    val result = inorderTraversal(root)
    println(result) // Output: [4, 2, 5, 1, 6, 3, 7]
}